freshmaker/pyxis.py

```diff
@@ -0,0 +1,205 @@
+
+
+class PyxisRequestError(Exception):
+    """
+    Error return as a response from Pyxis
+    """
+
+    def __init__(self, status_code, error_response):
+        """
+        Initialize Pyxis request error
+
+        :param int status_code: response status code
+        :param str or dict error_response: response content returned from Pyxis
+        """
+
+        self._status_code = status_code
+        self._raw = error_response
+
+    @property
+    def raw(self):
+        return self._raw
+
+    @property
+    def status_code(self):
+        return self._status_code
+
+    def __str__(self):
+        return 'Errors ({}):\n{}'.format(
+            len(self._raw['errors']),
+            '\n'.join(('    {}'.format(err['msg'])
+                      for err in self._raw['errors']))
+        )
+
+
+class Pyxis(object):
+    """ Interface for querying Pyxis"""
+
+    def __init__(self, server_url):
+        self._server_url = server_url.rstrip('/')
+        # add api version to root url
+        self._api_root = f"{self._server_url}/v1"
+        # list of default fields to include in request to pyxis
+        self._default_include_fields = ['page_size', 'total']
+
+    def _make_request(self, entity, params):
+        """ Send a request to Pyxis
+
+        :param str entity: entity part to construct a full URL for request.
+        :param dict params: Pyxis query parameters.
+        :return: Json response from Pyxis
+        :rtype: dict
+        :raises PyxisRequestError: If Pyxis returns error as a response
+        """
+        entity_url = '{}/{}'.format(self._api_root, entity)
+        default_params = {"page_size": "100"}
+        default_params.update(params)
+
+        response = requests.get(entity_url, params=default_params,
+                                auth=HTTPKerberosAuth(mutual_authentication=OPTIONAL))
+
+        if response.status_code == HTTPStatus.OK:
+            return response.json()
+
+        # Warn early, in case there is an error in the error handling code below
+        log.warning("Request to %s gave %r", response.request.url, response)
+
+        raise PyxisRequestError(response.status_code, response.json())
+
+    def _pagination(self, entity, params):
+        """
+        Process all pages in Pyxis
+
+        :param str entity: what data/entity to request from Pyxis
+        :param dict params: parameters to add to GET request
+        :return: list of all 'data' fields from responses from Pyxis
+        :rtype: list
+        """
+        ret = []
+        response_data = self._make_request(entity, params=params)
+        ret.extend(response_data.get('data', []))
+        page = 1
+        last_page = ceil(response_data['total'] / response_data['page_size'])
+        while page < last_page:
+            params.update({"page": page})
+            response_data = self._make_request(entity, params=params)
+            ret.extend(response_data.get('data', []))
+            page += 1
+        return ret
```

Comment:
Optional: This is more concise (doesn't repeat code) and addresses the previous comment (untested).
```suggestion
        request_params = copy.deepcopy(params)
        ret = []
        page = 0
        while True:
            request_params["page"] = page
            response_data = self._make_request(entity, params=request_params)
            # When the page after the actual last page is released, data will be an empty list
            if not response_data['data']:
                break
            ret.extend(response_data['data'])
            page += 1
            
        return ret
```